gui/tiletypes.lua

```diff
@@ -0,0 +1,1667 @@
+BoxSelection = defclass(BoxSelection, widgets.Window)
+BoxSelection.ATTRS {
+    box=DEFAULT_NIL, -- For output
+    screen=DEFAULT_NIL, -- Allows the DimensionsTooltip to be shown
+    tile_map=TILE_MAP,
+    avoid_rect=DEFAULT_NIL,
+    on_confirm=DEFAULT_NIL,
+    first_point=DEFAULT_NIL,
+    last_point=DEFAULT_NIL,
+    flat=false,
+    ascii_fill=false,
+}
+
+function BoxSelection:init()
+    self.frame = { w=0, h=0 }
+
+    -- Set the cursor to the center of the screen
+    local dims = dfhack.gui.getDwarfmodeViewDims()
+    guidm.setCursorPos {
+        x= df.global.window_x + (dims.map_x2 - dims.map_x1 + 1) // 2,
+        y= df.global.window_y + (dims.map_y2 - dims.map_y1 + 1) // 2,
+        z= df.global.window_z,
+    }
+
+    -- Show cursor
+    df.global.game.main_interface.main_designation_selected = df.main_designation_type.TOGGLE_ENGRAVING -- Alternative: df.main_designation_type.REMOVE_CONSTRUCTION
+    self.lastCursorPos = guidm.getCursorPos()
+
+    if self.screen and self.screen.addviews then
+        self.screen:addviews{
+            widgets.DimensionsTooltip{
+                view_id="dimensions_tooltip",
+                get_anchor_pos_fn=function()
+                    if self.first_point and self.flat then
+                        return { x= self.first_point.x, y= self.first_point.y, z= df.global.window_z }
+                    end
+                    return self.first_point
+                end,
+            },
+        }
+    else
+        qerror("No screen provided to BoxSelection, unable to display DimensionsTooltip")
+    end
+end
+
+function BoxSelection:confirm()
+    if self.first_point and self.last_point
+        and dfhack.maps.isValidTilePos(self.first_point.x or -1, self.first_point.y or -1, self.first_point.z or -1)
+        and dfhack.maps.isValidTilePos(self.last_point.x or -1, self.last_point.y or -1, self.last_point.z or -1)
+    then
+        self.box = Box{
+            self.first_point,
+            self.last_point
+        }
+        if type(self.on_confirm) == "function" then
+            self.on_confirm(self.box)
+        end
+    end
+end
+
+function BoxSelection:clear()
+    self.box = nil
+    self.first_point = nil
+    self.last_point = nil
+end
+
+function BoxSelection:onInput(keys)
+    if BoxSelection.super.onInput(self, keys) then
+        return true
+    end
+    if keys.LEAVESCREEN or keys._MOUSE_R then
+        return false
+    end
+
+    local mousePos = dfhack.gui.getMousePos(true)
+    local cursorPos = copyall(df.global.cursor)
+
+    if keys.SELECT then
+        if self.first_point and not self.last_point then
+            if not self.flat or cursorPos.z == self.first_point.z then
+                df.global.cursor.x = math.max(math.min(cursorPos.x, df.global.world.map.x_count - 1), 0)
+                df.global.cursor.y = math.max(math.min(cursorPos.y, df.global.world.map.y_count - 1), 0)
+                cursorPos = guidm.getCursorPos()
```

Comment:
This code was meant to force the keyboard cursor to stay inside the map bounds, but I guess following standard functionality does make sense.
Might go back later and add a flag that forces the cursor to stay in the map though, since I do find that base funcionality kinda annoying.